synapse/lib/remcycle.py

```diff
@@ -0,0 +1,876 @@
+
+    ::
+
+        {
+          "api": [
+            [
+              "someplace"
+            ],
+            {
+              "domore": 0
+            }
+          ],
+          "doc": "api example",
+          "http": {
+            "headers": {
+              "token-goodness": "sekrit token"
+            }
+          },
+          "ingests": {
+            "ingest_definition": {
+              "ingest": {
+                "forms": [
+                  [
+                    "inet:ipv4",
+                    {
+                      "var": "ip"
+                    }
+                  ]
+                ],
+                "vars": [
+                  [
+                    "ip",
+                    {
+                      "path": "ip"
+                    }
+                  ]
+                ]
+              }
+            }
+          },
+          "url": "http://vertex.link/api/v4/geoloc/{{someplace}}/info?domore={{domore}}&apikey={APIKEY}",
+          "vars": [
+            ["APIKEY", "8675309"]
+          ]
+        }
+
+    ::
+
+    This example should be interpreted as the following:
+
+        1. The APIKEY value in the 'vars' will be set in the URL, resulting in
+           the following default url:
+
+        ::
+
+            "http://vertex.link/api/v4/geoloc/{someplace}/info?domore={domore}&apikey=8675309"
+
+        ::
+
+        2. The HTTP request will have the header "token-goodness" set to
+           "sekrit token" for the request.
+        3. The caller must provide the "someplace" value in the api_args when
+           calling build_http_request.  The caller may provide the "domore"
+           value if they want to override the default value of "0".
+        4. An Ingest will be created for parsing the data from the API and
+           made available to the Hypnos object.
+
+    :param api_config: API Endpoint configuration outlined above.
+    :param namespace_http_config: Default HTTPRequent configuration values.
+    """
+
+    def __init__(self, api_config, namespace_http_config=None):
+        if namespace_http_config:
+            self.namespace_http_config = namespace_http_config.copy()
+        else:
+            self.namespace_http_config = {}
+        validate_http_values(self.namespace_http_config)
+        self._raw_config = dict(api_config)
+        self.required_keys = ['url',
+                              'doc'
+                              ]
+        self.optional_keys = ['vars',
+                              'http',
+                              'api',
+                              'ingests'
+                              ]
+        self.doc = ''
+        self.url_template = ''
+        self.url_vars = {}
+        self.effective_url = ''
+        self.request_defaults = {}
+        self.api_args = []
+        self.api_kwargs = {}
+        self.gests = {}
+
+        self._parse_config()
+        self._default_client_config = self._build_default_req_dict()
+
+    def _parse_config(self):
```

Comment:
this seems like a candidate for functionality to merge directly into Configable  ( ie, make a config option required and add an reqValidConf() method to Configable ( which checks that all required options are set )  That also lets us potentially do type casting ( and doc helpering ) all in one place.